src/test/java/se/liu/ida/hefquin/engine/data/impl/VocabularyMappingTest.java

```diff
@@ -74,6 +74,226 @@ public void TranslateTriplePatternTest() {
 		assertEquals(testData.object2, translationTriples);
 	}
 	
+	@Test
+	public void TranslateSubjectTest() {
+		final Set<Triple> testTriples = new HashSet<>();
+		//Equality
+		Node s = NodeFactory.createLiteral("s1");
+		Node p = OWL.sameAs.asNode();
+		final Node o1 = NodeFactory.createLiteral("s2");
+		testTriples.add(new Triple(s, p, o1));
+				
+		//Multiple mappings for same subject
+		final Node o2 = NodeFactory.createLiteral("s3");
+		testTriples.add(new Triple(s, p, o2));
+		
+		final VocabularyMapping vm = new VocabularyMappingImpl(testTriples);
+		
+		p = NodeFactory.createLiteral("p");
+		final Node o = NodeFactory.createLiteral("o");
+		final TriplePattern testTp = new TriplePatternImpl(s, p, o);
+		final SPARQLGraphPattern translation = vm.translateTriplePattern(testTp); 
+		final Set<Triple> translationTriples = new HashSet<>();
+		assertTrue(translation instanceof SPARQLUnionPatternImpl);
+		for (final SPARQLGraphPattern i : ((SPARQLUnionPatternImpl) translation).getSubPatterns()) {
+			assertTrue(i instanceof TriplePattern);
+			translationTriples.add(((TriplePattern) i).asJenaTriple());
+		}
+			
+		final Set<Triple> expectedResults = new HashSet<>();
+		expectedResults.add(new Triple(o1, p, o));
+		expectedResults.add(new Triple(o2, p, o));
+		
+		assertEquals(translationTriples, expectedResults);
+	}
+	
+	@Test
+	public void TranslatePredicateTest() {
+		final Set<Triple> testTriples = new HashSet<>();
+		//Equality
+		Node s = NodeFactory.createLiteral("p1");
+		Node p = OWL.equivalentProperty.asNode();
+		final Node t1 = NodeFactory.createLiteral("p2");
+		testTriples.add(new Triple(s, p, t1));
+				
+		//Predicate inverse
+		p = OWL.inverseOf.asNode();
+		final Node t2 = NodeFactory.createLiteral("Not p1");
+		testTriples.add(new Triple(s, p, t2));
+			
+		//Predicate subProperty
+		final Node t3 = NodeFactory.createLiteral("Subtype");
+		p = RDFS.subPropertyOf.asNode();
+		testTriples.add(new Triple(t3, p, s));
+		
+		//Predicate Intersection
+		p = OWL.equivalentProperty.asNode();
+		final Node t4 = NodeFactory.createBlankNode();
+		testTriples.add(new Triple(s, p, t4));	
+		p = OWL.intersectionOf.asNode();
+		final Node t5 = NodeFactory.createLiteral("p3");
+		testTriples.add(new Triple(t4, p, t5));		
+		final Node t6 = NodeFactory.createLiteral("p4");
+		testTriples.add(new Triple(t4, p, t6));
+		
+		//Predicate Union
+		p = OWL.equivalentProperty.asNode();
+		final Node t7 = NodeFactory.createBlankNode();
+		testTriples.add(new Triple(s, p, t7));	
+		p = OWL.unionOf.asNode();
+		final Node t8 = NodeFactory.createLiteral("p5");
+		testTriples.add(new Triple(t7, p, t8));		
+		final Node t9 = NodeFactory.createLiteral("p6");
+		testTriples.add(new Triple(t7, p, t9));
+
+		final VocabularyMapping vm = new VocabularyMappingImpl(testTriples);
+		
+		p = s;
+		s = NodeFactory.createLiteral("s");
```

Comment:
Is there really a need to reuse these variables like this? it seems wrong to me to assign s (subject) into p (predicate). Please just use fresh variables every time.

Reply:
I did this because I thought I needed to use the same node in order for the results to equate, but maybe it isn't the case? I will try to just make new ones